src/goFormat.ts

```diff
@@ -23,30 +24,37 @@ export class GoDocumentFormattingEditProvider implements vscode.DocumentFormatti
 		let formatTool = goConfig['formatTool'] || 'goreturns';
 		let formatFlags = goConfig['formatFlags'].slice() || [];
 
-		// We ignore the -w flag that updates file on disk because that would break undo feature
-		if (formatFlags.indexOf('-w') > -1) {
-			formatFlags.splice(formatFlags.indexOf('-w'), 1);
-		}
-
-		// Fix for https://github.com/Microsoft/vscode-go/issues/613 and https://github.com/Microsoft/vscode-go/issues/630
-		if (formatTool === 'goimports' || formatTool === 'goreturns') {
-			formatFlags.push('-srcdir', filename);
-		}
-
-		// Since goformat supports the style flag, set tabsize if user has not passed any flags
-		if (formatTool === 'goformat' && formatFlags.length === 0 && options.insertSpaces) {
-			formatFlags.push('-style=indent=' + options.tabSize);
-		}
+		isModSupported(document.uri).then(isMod => {
+			formatTool = 'goimports';
```

Comment:
both the issue and the description of your pull request say:

> use `goimports` if user's choice is `goreturns` in module mode

But this will effectivly overwrite formatTool also if the value is `gofmt`.

It's also unnecessary to call `isModSupported` if the `formatTool` is already `goimports`.

Reply:
@lloiser I'm not sure how do you intend to invoke `isModSupported` as per the above mentioned case.
Do you mean something like checking for whether `formatTool` is `goimports` and invoking the method appropriately :thinking: 
```js
if (formatTool === 'goimports') {
 //
}
```

I'm not familiar with the codebase as mentioned before. Kindly suggest the required changes :smile:

Reply:
If the chosen tool is either gofmt or goimports, then no change is needed.
Else, you need to invoke `isModSupported` to figure out if modules are being used or not.
Something like the below

```typescript
private getFormatTool(goConfig: vscode.WorkspaceConfiguration, fileUri: vscode.Uri): Thenable<string> {
	let formatTool = goConfig["formatTool"] || "goreturns";
	if (formatTool === "gofmt" || formatTool === "goimports") {
		return Promise.resolve(formatTool);
	}
	return isModSupported(fileUri).then(isMod => {
		return isMod ? "goimports": formatTool;
	})
} 
```

On second thoughts, I wonder if we should choose to use `gofmt` instead. `gofmt` is installed by default for all, whereas one has to specifically install `goimports`. Also, older versions of `goimports` don't support modules, so `goimports` has to be installed, meaning we need to prompt the user to do so.

Regardless, we are choosing to use a different tool than what the user meant to use. Therefore, we should notify the user of this change, and update workspace settings so that we don't have to make this check every time the file needs to be formatted.

For reference, this updating of workspace settings when using modules is already being done for the `go.infergopath` setting  at https://github.com/Microsoft/vscode-go/blob/0.9.2/src/goModules.ts#L55. We can update the same

Reply:
@ramya-rao-a Is this the way of implementation that you intended :thinking: